### ISSUE TITLE
Is it good to handle multiple requests in a single handler?

### ISSUE THREAD

**Issue by lecoque** [jbogard/MediatR#281]
Let's take some CRUD application for example. I have 4 requests for Create, Read, Update, Delete actually 5 with Get All (I'm building resource-oriented web API)

Is it good to handle them in a single handler like or is something wrong with this approach?:

```csharp
public class ProductRequestHandler : 
    IRequestHandler<CreateProductRequest, ProductResult>, 
    IRequestHandler<GetProductRequest, ProductResult>
    IRequestHandler<GetAllProductsRequest, ProductListResult>
    IRequestHandler<UpdateProductRequest, ProductResult>
    IRequestHandler<DeleteProductRequest, bool>
{
...
}
```

I'm trying to move away from DDD services that handle similar logic but I don't know if it's worth it.

Also, could someone tell me why shouldn't my Create request return anything if I wanted to apply CQRS? What's the benefit of commands not returning anything? Doing this just for the sake of "responsibility segregation" doesn't convince me. My DDD services returned stuff in my Create methods and my Update methods.

I'm sorry, the question might be trivial for some of you but it's actually pretty confusing topic. And I'd like to get my mind right

**Comment by no1melman** [comment]
TL:DR; If you implement all those interfaces in that single class - are you not just going back to the n-tier architecture days, where the service layer was a class that held all the logic. If you want to do that, I don't see the point in using MediatR.

Longer version

MediatR allows you to break your application up into distinct requests (like microservices) and handle a specific pipeline for each. This gives enormous benefits, both in maintainability and flexibility.

I feel that with your approach you're kinda of breaking those principles thus rendering the underlying framework a tad irrelevant.

**Comment by lecoque** [comment]
@no1melman that's something I was thinking. I felt something was wrong with the code and that type of handling isn't much different than having services.

Could you give some real-world example of such benefits based on experiences?

It would be nice to get some response from @jbogard too

**Comment by jbogard** [comment]
Ditto. Don't combine your handlers, keep them separate, reduce coupling across handlers. For a real example, you can look at my various ContosoUniversity projects under my root profile.

**Comment by SoftInvent** [comment]
A comment just after this closed.

I will say that it depends. I totally agree with Bogard that, the way you illustrate it in you opening comment is not the way to go.
However, I've just implemented a scenario where I have a command "CreateUserCommand" which needs to know whether the user is already a customer or not. The way we do this is by looking up insurance coverages for the user's danish SSN and if any exist, the user is already a customer. 
However, this is quite an expensive process and we do snapshotting and caching of this to improve performance. Furthermore this lookup is done in several places in our API. 
So, we have implemented a CustomerCoveragesQuery, to handle this. It goes through the pipeline and is automatically snapshotted (is that a word?) and cached. This is done by our generic Snapshot- and CachingBehaviors.
Now, back to the question. Our CreateUserCommandHandler actually fires off a CustomerCoveragesQuery and reacts on the result but now with fully transparent caching and snapshotting.
Agreed, this give a coupling between the command and query and could have been solved by a simple service layer, but we wanted to stick with CQRS as much as possible.
Cheers :-)

**Comment by lecoque** [comment]
@SoftInvent so you're injecting `IMediator` to your handlers and sending queries, that seems valid when using full CQRS.

**Comment by SoftInvent** [comment]
@lecoque Yes, I've create an abstract CommandHandler<TCommand, TResult> and QueryHandler<TQuery, TResult> which both gets an IMediator injected, so they can either send new commands or queries or publish notifications.

**Comment by andresmolivares** [comment]
If you practice proper SOA, you would not need CQRS, let alone mediatR abstraction. What is the need to decouple the controller from the service? Are you just coupling the controller to mediatR library, unnecessarily I might add, since it is not resolving any issues...as mentioned that's what code reviews are for. If I normally go Controller -> Service -> Repository, now I some how benefit but going Controller -> MediatR -> Handler -> Service -> Repository or whatever variation workflow is required to implement MediatR, how is this case better? The previous is direct. Additionally, If for whatever reason I needed to go CQRS, what is the Send method doing differently that I can physically leverage than the handler.Handle?

I can understand the intent behind the pattern, but it's not fixing anything beside adding rigor to "dirty" SOA implementations, or else why would any of this be a discussion.

I agree with all in this post: https://stackoverflow.com/questions/50834935/replacing-service-layer-with-mediatr-is-it-worth-to-do-it

**Comment by andresmolivares** [comment]
Moreso, when I look at what Send is doing in the MediatR implementation:

`public Task<TResponse> Send<TResponse>(IRequest<TResponse> request, CancellationToken cancellationToken = default)
    {
        if (request == null)
        {
            throw new ArgumentNullException(nameof(request));
        }

        var requestType = request.GetType();

        var handler = (RequestHandlerWrapper<TResponse>)_requestHandlers.GetOrAdd(requestType,
            static t => (RequestHandlerBase)(Activator.CreateInstance(typeof(RequestHandlerWrapperImpl<,>).MakeGenericType(t, typeof(TResponse)))
                                             ?? throw new InvalidOperationException($"Could not create wrapper type for {t}")));

        return handler.Handle(request, cancellationToken, _serviceFactory);
    }`
    
   why would I ever need to Activator.CreateInstance of anything, it's a request for crying out loud. WHAT....why are we really doing this? C'mon people.

**Comment by andresmolivares** [comment]
> Could you give some real-world example of such benefits based on your experiences?



> A comment just after this closed.
> 
> I will say that it depends. I totally agree with Bogard that, the way you illustrate it in you opening comment is not the way to go. However, I've just implemented a scenario where I have a command "CreateUserCommand" which needs to know whether the user is already a customer or not. The way we do this is by looking up insurance coverages for the user's danish SSN and if any exist, the user is already a customer. However, this is quite an expensive process and we do snapshotting and caching of this to improve performance. Furthermore this lookup is done in several places in our API. So, we have implemented a CustomerCoveragesQuery, to handle this. It goes through the pipeline and is automatically snapshotted (is that a word?) and cached. This is done by our generic Snapshot- and CachingBehaviors. Now, back to the question. Our CreateUserCommandHandler actually fires off a CustomerCoveragesQuery and reacts on the result but now with fully transparent caching and snapshotting. Agreed, this gives a coupling between the command and query and could have been solved by a simple service layer, but we wanted to stick with CQRS as much as possible. Cheers :-)

So you're admitting that SOA could have easily been implemented for the described use case, you know hae coupleing between Handler and your Query and you still want to sticj to CQRS for yet still an unexplained and baseless reason. Am I undersatnding your response correctly?

**Comment by bugproof** [comment]
> > Could you give some real-world example of such benefits based on your experiences?
> 
> > A comment just after this closed.
> > I will say that it depends. I totally agree with Bogard that, the way you illustrate it in you opening comment is not the way to go. However, I've just implemented a scenario where I have a command "CreateUserCommand" which needs to know whether the user is already a customer or not. The way we do this is by looking up insurance coverages for the user's danish SSN and if any exist, the user is already a customer. However, this is quite an expensive process and we do snapshotting and caching of this to improve performance. Furthermore this lookup is done in several places in our API. So, we have implemented a CustomerCoveragesQuery, to handle this. It goes through the pipeline and is automatically snapshotted (is that a word?) and cached. This is done by our generic Snapshot- and CachingBehaviors. Now, back to the question. Our CreateUserCommandHandler actually fires off a CustomerCoveragesQuery and reacts on the result but now with fully transparent caching and snapshotting. Agreed, this gives a coupling between the command and query and could have been solved by a simple service layer, but we wanted to stick with CQRS as much as possible. Cheers :-)
> 
> So you're admitting that SOA could have easily been implemented for the described use case, you know have coupling between your Handler and Query and you still want to stick to CQRS for a still yet unexplained, and seemingly baseless, reason. Am I understanding your response correctly?

I wrote that comment in 2018. I stopped bothering as much with patterns let alone MediatR library which is just an event handling implementation with a fancy name.

**Comment by jbogard** [comment]
> If you practice proper SOA, you would not need CQRS, let alone mediatR abstraction. What is the need to decouple the controller from the service? Are you just coupling the controller to mediatR library, unnecessarily I might add, since it is not resolving any issues...as mentioned that's what code reviews are for. If I normally go Controller -> Service -> Repository, now I some how benefit but going Controller -> MediatR -> Handler -> Service -> Repository or whatever variation workflow is required to implement MediatR, how is this case better? The previous is direct. Additionally, If for whatever reason I needed to go CQRS, what is the Send method doing differently that I can physically leverage than the handler.Handle?
> 
> I can understand the intent behind the pattern, but it's not fixing anything beside adding rigor to "dirty" SOA implementations, or else why would any of this be a discussion.
> 
> I agree with all in this post: https://stackoverflow.com/questions/50834935/replacing-service-layer-with-mediatr-is-it-worth-to-do-it

Um...CQRS has nothing to do with SOA. They are completely orthogonal concepts/architectures.

And the flow is slightly different, it's going from: Controller -> Service -> Repository to Controller -> Handler. I eliminate the two redundant abstractions, service and repository. If you decide to keep those additional layers then yes the handler gives you nothing. But I don't have those extraneous layers because once I encapsulate my logic in a handler, those extra layers of indirection don't add any value.

**Comment by jbogard** [comment]
I wasn't implying I don't use MediatR, I still do. I combine MediatR with CQRS as the handlers and requests let us create the "two objects where once there was one".

CQRS is not about direction or indirection, it's recognizing that Reads and Writes have vastly different heuristics and combining these into common objects (services, models) leads to conflicts. Comparing I think what you're referring to as "n-tier" architecture (Controller -> Service -> Data Tier) is a separate thing altogether, I've done a few talks about that (look around for SOLID architecture in slices not layers).

**Comment by andresmolivares** [comment]
Understood Jimmy,

I was actually getting introduced to CQRS and MediatR simultaneously and was trying to understand the value of MediatR. Naming convention is misleading but it is a way of implementing CQRS separation and plugs into the pipeline nicely. I essentially could just do basic CQRS, but even that is just a flavor of SOA...any really only guidance because there is nothing really stopping by from a Command executing a GET or a Query doing a PUT, POST or delete.

Thanks for the follow up,

Andres

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jbogard** [comment]
Pull that code into a common class, it's no different than any other duplicate code.
